### PR TITLE
drivers: ieee802154: cc13xx_cc26xx: sub-ghz support

### LINKED ISSUE #26315: ieee802154: cc13xx_cc26xx: sub_ghz support
**Is your enhancement proposal related to a problem? Please describe.**
The ieee802154 cc13xx_cc26xx could probably also use an option for Sub GHz support as well. This might be a little easier to do once #26312 is done as well, since the rf api is (somewhat?) common.

It would be helpful to also have a choice in Kconfig or DT for the radio frequency.

Edit: According to the [TRM](https://www.ti.com/lit/pdf/swcu185) section 25.10.5.2.1, IEEE 802.15.4g (sub ghz) mode is supported via the "proprietary" radio. It does not seem that the IEEE mode radio commands are used at all and that the `PROP` commands are used for setup, TX, RX, and CS.

**Describe the solution you'd like**
After #26312 is done, I could probably take a crack at this.

**Describe alternatives you've considered**

**Additional context**

Depends on #26312

### PR DESCRIPTION
**Pull request by cfriedt**
This change adds IEEE 802.15.4g (Sub GHz) support for the cc1352r.

The 2.4 GHz radio and the Sub GHz radio are capable of operating simultaneously.

Fixes #26315

### CODE REVIEW

**Comment by cfriedt** [comment]
Notes:
* will likely require a small fixup in `power.c` and possibly some `Kconfig` files once #26316 is merged.
* will need to undo edits to some of the sample applications as well, that I had originally hard-coded to demonstrate SubGHz support
* Also note, this is 802.15.4g and does not provide the BPSK PHY - just the SUN 2-FSK PHY.

**Comment by cfriedt** [comment]
Furthermore, it requires zephyrproject-rtos/hal_ti#22 in order to pull in the multi-protocol RF backend.

Will likely also require some whitespace corrections

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+	uint8_t mac[8];
+
+	struct k_mutex tx_mutex;
+
+	dataQueue_t rx_queue;
+	rfc_dataEntryPointer_t rx_entry[CC13XX_CC26XX_NUM_RX_BUF];
+	uint8_t rx_data[CC13XX_CC26XX_NUM_RX_BUF][CC13XX_CC26XX_RX_BUF_SIZE] __aligned(4);
+
+	uint8_t tx_data[CC13XX_CC26XX_TX_BUF_SIZE] __aligned(4);
```
Comment (jukkar) [inline drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h:85]: Is aligned really needed here as this struct is not packed? The fields will be aligned by compiler anyway.

File: include/net/ieee802154_radio.h [hunk 2]
```diff
@@ -209,10 +209,8 @@ struct ieee802154_radio_api {
 			 enum ieee802154_config_type type,
 			 const struct ieee802154_config *config);
 
-#ifdef CONFIG_NET_L2_IEEE802154_SUB_GHZ
```
Comment (jukkar) [inline include/net/ieee802154_radio.h:212]: Why do we remove this?

File: samples/hello_world/prj.conf [hunk 3]
```diff
@@ -1 +1,30 @@
...
+CONFIG_INIT_STACKS=y
+
+# These were both 1024 by default, and that was overflown fairly quickly
+# Need to measure stack sizes for all threads that are created
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+
+# Power management
+CONFIG_SYS_POWER_MANAGEMENT=y
+CONFIG_SYS_POWER_SLEEP_STATES=y
```
Comment (jukkar) [inline samples/hello_world/prj.conf:30]: Hmm, I do not think we should add any networking config to hello_world application.

File: samples/net/sockets/echo_client/overlay-802154.conf [hunk 4]
```diff
@@ -15,4 +15,10 @@ CONFIG_NET_L2_IEEE802154=y
...
 CONFIG_NET_L2_IEEE802154_LOG_LEVEL_INF=y
 
-CONFIG_NET_CONFIG_IEEE802154_CHANNEL=26
+#CONFIG_NET_CONFIG_IEEE802154_CHANNEL=26
+CONFIG_IEEE802154_CC13XX_CC26XX=n
+#CONFIG_NET_CONFIG_IEEE802154_DEV_NAME="IEEE802154_0"
+
+CONFIG_NET_CONFIG_IEEE802154_CHANNEL=1
+CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ=y
+CONFIG_NET_CONFIG_IEEE802154_DEV_NAME="IEEE802154_1"
```
Comment (jukkar) [inline samples/net/sockets/echo_client/overlay-802154.conf:24]: This does not look right. Perhaps provide separate overlay for sub ghz config instead. Same comment to changes in other samples.

**Review by jukkar — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c [hunk 5]
```diff
@@ -0,0 +1,902 @@
...
+static void ieee802154_cc13xx_cc26xx_subg_setup_rx_buffers(
+	struct ieee802154_cc13xx_cc26xx_subg_data *drv_data);
+
+DEVICE_DECLARE(ieee802154_cc13xx_cc26xx_subg);
+
+/* Overrides from SmartRF Studio 7 2.18.0 */
+static uint32_t overrides_sub_ghz[] = {
+	/* DC/DC regulator: In Tx, use DCDCCTL5[3:0]=0x7 (DITHER_EN=0 and IPEAK=7). */
+	(uint32_t)0x00F788D3,
+	// override_prop_common_sub1g.xml
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:49]: I haven't yet run it through checkpatch.pl - sorry! Will fix up nits & whitespace now.

File: include/net/ieee802154_radio.h (hunk 2 above)
Comment (cfriedt) [reply include/net/ieee802154_radio.h:212]: This device supports `CONFIG_NET_L2_IEEE802154` and `CONFIG_NET_L2_IEEE802154_SUB_GHZ` simultaneously. Also, if `CONFIG_IEEE802154_RAW_MODE` is set, then neither `CONFIG_NET_L2_IEEE802154` or `CONFIG_NET_L2_IEEE802154_SUB_GHZ` are enabled, which results in 2.4 GHz radios being able to change their channels but not SubGHz radios.

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h (hunk 1 above)
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h:85]: I'll give it a shot. The original 2.4 GHz driver aligned buffers that are passed to the radio.

**Comment by cfriedt** [comment]
Until #26316 is merged, SubGHz and 2.4 GHz cannot be used simultaneously, so to test this, please use the following:

server:
```bash
west build \
    -p always \
    -b cc1352r1_launchxl \
    samples/net/sockets/echo_server \
    -- \
    -DOVERLAY_CONFIG=overlay-802154-subg.conf \
    -DCONFIG_NET_CONFIG_IEEE802154_DEV_NAME="\"IEEE802154_1\"" \
    -DCONFIG_IEEE802154_CC13XX_CC26XX=n \
    -DCONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ=y
```

client:
```bash
west build \
    -p always \
    -b cc1352r1_launchxl \
    samples/net/sockets/echo_client \
    -- \
    -DOVERLAY_CONFIG=overlay-802154-subg.conf \
    -DCONFIG_NET_CONFIG_IEEE802154_DEV_NAME="\"IEEE802154_1\"" \
    -DCONFIG_IEEE802154_CC13XX_CC26XX=n \
    -DCONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ=y
```

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h (hunk 1 above)
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h]: It's not needed

**Comment by cfriedt** [comment]
So odd that checkpatch.pl doesn't catch some long lines. I don't understand... was the 80-character limit silently extended?

**Review by vanti — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+
+	r = ieee802154_cc13xx_cc26xx_subg_channel_to_frequency(channel, &freq, &fract);
+	if (r < 0) {
+		return -EINVAL;
+	}
+
+	/* Abort FG and BG processes */
+	if (ieee802154_cc13xx_cc26xx_subg_stop(dev) < 0) {
+		r = -EIO;
+		goto out;
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:237]: Yes! Thanks - that was originally below the `k_mutex_lock()`, but I moved it up and clearly forgot to just return! Good cactch!

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+					const struct ieee802154_config *config)
+{
+	return -ENOTSUP;
+}
+
+uint16_t ieee802154_cc13xx_cc26xx_subg_get_subg_channel_count(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 11;
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:483]: add a comment about this number

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c [hunk 8]
```diff
@@ -0,0 +1,705 @@
...
+	{ 9, RF_TxPowerTable_DEFAULT_PA_ENTRY(28, 3, 0, 31) },
+	{ 10, RF_TxPowerTable_DEFAULT_PA_ENTRY(18, 2, 0, 31) },
+	{ 11, RF_TxPowerTable_DEFAULT_PA_ENTRY(26, 2, 0, 51) },
+	{ 12, RF_TxPowerTable_DEFAULT_PA_ENTRY(16, 0, 0, 82) },
+	{ 13, RF_TxPowerTable_DEFAULT_PA_ENTRY(36, 0, 0, 89) },
+	{ 14, RF_TxPowerTable_DEFAULT_PA_ENTRY(63, 0, 1, 0) },
+	RF_TxPowerTable_TERMINATION_ENTRY
+};
+
+static inline int ieee802154_cc13xx_cc26xx_subg_channel_to_frequency(uint16_t channel, uint16_t *frequency, uint16_t *fractFreq)
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:83]: don't use camel case for your own code (one coming from the HAL is obviously fine).

Btw, you seem to follow the 80 chars limits properly almost everywhere but it in some places like this one. Please keep consistency all over the code.

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+static int ieee802154_cc13xx_cc26xx_subg_stop(
+	const struct device *dev);
+static int ieee802154_cc13xx_cc26xx_subg_rx(
+	const struct device *dev);
+static void ieee802154_cc13xx_cc26xx_subg_setup_rx_buffers(
+	struct ieee802154_cc13xx_cc26xx_subg_data *drv_data);
+
+DEVICE_DECLARE(ieee802154_cc13xx_cc26xx_subg);
+
+/* Overrides from SmartRF Studio 7 2.18.0 */
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:45]: a bit more explanation would be welcomed here

File: include/net/ieee802154_radio.h (hunk 2 above)
Comment (tbursztyka) [reply include/net/ieee802154_radio.h:212]: fine by me

**Review by tbursztyka — APPROVED** [review]
small changes requested

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c (hunk 7 above)
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:483]: Jeez - yeah, that one totally got by me. Good catch!

File: drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h [hunk 10]
```diff
@@ -0,0 +1,100 @@
+/*
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h:1]: Would you be against pushing that back for a later ticket?

[the remark this answers is not in the document]
Comment (vanti) [reply drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.h:1]: I am ok with that. Just fix the description for "menuconfig IEEE802154_CC13XX_CC26XX_SUB_GHZ" and I can approve.

**Comment by cfriedt** [comment]
@vanti - I've defaulted `IEEE802154_CC13XX_CC26XX_SUB_GHZ` to `n' for 2 reasons:

1. We can check integration with zephyrproject-rtos/hal_ti#22 (required for SubGHz to run) and
1. There is still an issue with multiple concurrent network interfaces (described in #29750)

For the second item, we're seeing things like the following:
```
*** Booting Zephyr OS build zephyr-v2.4.0-1206-g98490c598ed7  ***
[00:00:00.001,983] <wrn> net_if: You have 1 IPv6 net_if addresses but 2 network interfaces
[00:00:00.002,014] <wrn> net_if: Consider increasing CONFIG_NET_IF_MAX_IPV6_COUNT value.
[00:00:00.010,650] <err> net_config: Cannot setup IEEE 802.15.4 interface (-22)
[00:00:00.010,650] <inf> net_config: Initializing network
[00:00:00.010,650] <inf> net_config: Waiting interface 1 (0x20001438) to be up...
```
Tried increasing `CONFIG_NET_IF_MAX_IPV6_COUNT`, but there is still no way to configure more than 1 IPv6 network interface. See #29750 for details.

**Review by vanti — APPROVED** [review]

**Comment by cfriedt** [comment]
Looks like the fix for this is in zephyrproject-rtos/hal_nxp#66

Is there anything I can do to sidestep that at the moment, or am I blocked until that fix is merged into hal/nxp & zephyr?

**Comment by MaureenHelm** [comment]
> Is there anything I can do to sidestep that at the moment, or am I blocked until that fix is merged into hal/nxp & zephyr?

It's been merged now, can you rebase?

**Comment by cfriedt** [comment]
Looks like someone has already filed an issue for [this](https://github.com/zephyrproject-rtos/zephyr/pull/29598/checks?check_run_id=1366668786) at #29853

**Review by jukkar — APPROVED** [review]

**Comment by akshaymishra** [comment]
Hello, does this support mean one can use zephyr 6lowpan with CC1310?

**Pull request merged**